### DISCUSSION TITLE
connect options config and scenarios

### DISCUSSION THREAD

**Discussion by joaqcid** [ngxs-labs/firestore-plugin#53] [Ideas]
> trackBy: **same for every action** or **not present**
> cancelPrevious: **false** (default value)
```
this.ngxsFirestoreConnect.connect(Action, {
  to: FirestoreStream,
});
```
```
this.ngxsFirestoreConnect.connect(Action, {
  to: FirestoreStream,
  trackBy: () => 'static',
  cancelPrevious: false
});
```

> trackBy: **same for every action** or **not present**
> cancelPrevious: **true**
```
this.ngxsFirestoreConnect.connect(Action, {
  to: FirestoreStream,
  cancelPrevious: true
});
```
```
this.ngxsFirestoreConnect.connect(Action, {
  to: FirestoreStream,
  trackBy: () => 'static',
  cancelPrevious: true
});
```

> trackBy: **different for every action**
> cancelPrevious: **false** (default value)
```
this.ngxsFirestoreConnect.connect(Action, {
  to: FirestoreStream,
  trackBy: (action) => action.payload // assumes payload is different on each dispatch,
});
```
```
this.ngxsFirestoreConnect.connect(Action, {
  to: FirestoreStream,
  trackBy: (action) => action.payload // assumes payload is different on each dispatch,
  cancelPrevious: false
});
```

> trackBy: **different for every action**
> cancelPrevious: **true**
```
this.ngxsFirestoreConnect.connect(Action, {
  to: FirestoreStream,
  trackBy: (action) => action.payload // assumes payload is different on each dispatch,
  cancelPrevious: true
});
```

-----
@markwhitfeld comments: 
Ok, so I'll summarise my discussions with @joaqcid in separate comments below.
Firstly, let's establish a baseline for what the connect options mean:
- `trackBy`: used to determine how many connections are triggered by the same action
  - setting a `trackBy` will allow for multiple connections, each of which is identified by the data returned by the `trackBy` that you provide
  - if you don't set a `trackBy` (the default) then there will only ever be one active connection for this action
  - similarly if you specify a constant value for the `trackBy` then there will only ever be one connection for this action
- `cancelPrevious`: used to determine how a new action will be handled
  - essentially this represents the choice of "Cancel previous connection" or "Ignore new connection"
  - if `cancelPrevious` is false (the default) then new actions are ignored if there is an existing connection with the same `trackBy`
  - if `cancelPrevious` is true, then the new action will be used to create a connection that will replace an existing connection with the same `trackBy`

The combination of these two properties fully determines how the `NgxsFireStoreConnect` handles its "connections" to streams. Obviously the source stream created by the action call can also have its own logic about when it completes.

Each combination of options is listed below:
- `{ trackBy: undefined, cancelPrevious: false, ... }` https://github.com/ngxs-labs/firestore-plugin/discussions/53#discussioncomment-2166749
- `{ trackBy: undefined, cancelPrevious: true, ... }` https://github.com/ngxs-labs/firestore-plugin/discussions/53#discussioncomment-2166760
- `{ trackBy: (a) => a.id, cancelPrevious: false, ... }` https://github.com/ngxs-labs/firestore-plugin/discussions/53#discussioncomment-2166776
- `{ trackBy: (a) => a.id, cancelPrevious: true, ... }` https://github.com/ngxs-labs/firestore-plugin/discussions/53#discussioncomment-2166793

**Comment by markwhitfeld** [comment]
[DEFAULT]
`{ trackBy: undefined, cancelPrevious: false, ... }`
Action: Get { id: string, other: string }

| # | Action | Payload | Connect | Connection Count | Closed By # |
| - | - | - | - | - | - |
| 1 | Get | `{ id: 1, other: 'a' }` | Yes | 1 | None |
| 2 | Get | `{ id: 2, other: 'a' }` | Ignored | 1 | - | 
| 3 | Get | `{ id:, other: 'a' }` | Ignored | 1 | - |
| 4 | Get | `{ id:, other: 'b' }` | Ignored | 1 | - |


Final connection count = 1
Actions connected after the sequence above:
`[ { id: 1, other: 'a' } ]`

PS. Equivalent to rxjs `exhaustMap`

**Comment by markwhitfeld** [comment]
`{ trackBy: undefined, cancelPrevious: true, ... }`
Action: Get { id: string, other: string }

Example sequence of actions and behavior:
| # | Action | Payload | Connect | Connections Count | Closed By # |
| - | - | - | - | - | - |
| 1 | Get | `{ id: 1, other: 'a' }` | Yes | 1 | 2 |
| 2 | Get | `{ id: 2, other: 'a' }` | Yes | 1 | 3 | 
| 3 | Get | `{ id: 3, other: 'a' }` | Yes | 1 | 4 |
| 4 | Get | `{ id: 1, other: 'b' }` | Yes | 1 | None |

Final connection count = 1
Actions connected after the sequence above:
`[ { id: 1, other: 'b' } ]`

PS. Equivalent to rxjs `switchMap`

**Comment by markwhitfeld** [comment]
`{ trackBy: (a) => a.id, cancelPrevious: false, ... }`
Action: Get { id: string, other: string }

Example sequence of actions and behavior:
| # | Action | Payload | Connect | Connections Count | Closed By # |
| - | - | - | - | - | - |
| 1 | Get | `{ id: 1, other: 'a' }` | Yes | 1 | None |
| 2 | Get | `{ id: 2, other: 'a' }` | Yes | 2 | None | 
| 3 | Get | `{ id: 3, other: 'a' }` | Yes | 3 | None |
| 4 | Get | `{ id: 1, other: 'b' }` | No | 3 | - |

Final connection count = 3
Actions connected after the sequence above:
`[ { id: 1, other: 'a' }, { id: 2, other: 'a' }, { id: 3, other: 'a' } ]`

PS. Equivalent to rxjs `groupBy(trackBy)` then `exhaustMap`
or rxjs `distinctUntilKeyChanged('id')` and then `mergeMap`

**Comment by markwhitfeld** [comment]
`{ trackBy: (a) => a.id, cancelPrevious: true, ... }`
Action: Get { id: string, other: string }

Example sequence of actions and behavior:
| # | Action | Payload | Connect | Connections Count | Closed By # |
| - | - | - | - | - | - |
| 1 | Get | `{ id: 1, other: 'a' }` | Yes | 1 | 4 |
| 2 | Get | `{ id: 2, other: 'a' }` | Yes | 2 | None | 
| 3 | Get | `{ id: 3, other: 'a' }` | Yes | 3 | None |
| 4 | Get | `{ id: 1, other: 'b' }` | Yes | 3 | None |

Final connection count = 3
Actions connected after the sequence above:
`[ { id: 1, other: 'b' }, { id: 2, other: 'a' }, { id: 3, other: 'a' } ]`

PS. Equivalent to rxjs `groupBy(trackBy)` then `switchMap`

**Comment by markwhitfeld** [comment]
Notes from discussion with @joaqcid around how to control concurrency....
- Group connections by an ID- track connection by ID
- how do we treat previous connections in the same group: `cancel` / `prefer`
- how do we handle connections from other groups?
  - `keep` / `close others` 
  - `maxConnections`: 1
  - `maxTrackedConnections`: 1
  - `keepPrevious`
  - `keepLast`
